xarray/plot/utils.py

```diff
@@ -98,6 +98,12 @@ def _build_discrete_cmap(cmap, levels, extend, filled):
     # copy the old cmap name, for easier testing
     new_cmap.name = getattr(cmap, "name", cmap)
 
+    # copy colors to use for bad, under, and over values in case they have been set to
+    # non-default values
+    new_cmap._rgba_bad = getattr(cmap, "_rgba_bad")
```

Comment:
this fails if `cmap` is a string (same for the two lines directly below). Try adding a default:
```suggestion
    new_cmap._rgba_bad = getattr(cmap, "_rgba_bad", None)
```

Reply:
Good point. I tested using `None` and it changes the default behaviour slightly, so I've added a fix getting the default values from `new_cmap._rgba_bad`, etc., instead. Then if `cmap` is a string, we continue to do exactly the same thing as was done before this PR.